### ISSUE TITLE
Orchard 2 should be better in its management of temporary content

### ISSUE THREAD

**Issue by CSurieux** [OrchardCMS/Orchard2#90]
In some situations, there is no displayed content, because in Orchard One's concept all content is permanently stored in DB.  So all ends with something packaged inside an hierachy of Shapes and displayed roughly in the Content Zone, with tricks everywhere. As an exemple click on a taxonomy term or a Tag. Projections could be an answer but they are permanently in DB.
There should be a structure for temporary and dynamics contents and benefits from all the zones/layers/placement.

**Comment by sebastienros** [comment]
I don't understand what temporary content is ;) sorry

**Comment by CSurieux** [comment]
A contentitem which will never go in DB, created on the fly  just for 1 request life time, built with temporary parts similar to 'permanent' ones but which don't store in DB. They dy with the sending of the result. For exemple when we click on a tag this is handled by a controler which builds a shape based on a list and pager, but there is nothing respecting a traditionnal display, we must use tricks to display something in zones other than Content. In this case the controler could create on the fly a contentitem with a temporary titlepart, temporary metadata, temporay taglistPart, temporary culture translation part, etc. Each temporay part will use it's existing driver to build display and use content management filters to produce  results. No special tricks, use the existing display flows and processes. Could go in cache but not in DB.

**Comment by bleroy** [comment]
I don't see why you couldn't do that. What did you try and how did it not work?

**Comment by CSurieux** [comment]
I don't see how to create something no enlisted in transaction. No standard calls. Are you saying that without complexe tricks I could do this? Why it is not used in tags, in search, etc ? I spent lot of time on tags to normalize my result screens. Only widget are working for this page.

**Comment by bleroy** [comment]
You can call New on ContentManager, and as long as you don't Create, it won't be persisted. In principle, that should work.

**Comment by CSurieux** [comment]
My idea would be to'normalize' this for all the case where the request does not end on a DB content but on something dynamic.

**Comment by CSurieux** [comment]
Relatednto #89

**Comment by bleroy** [comment]
What do you mean by "normalize"? What's abnormal about it? How is it related to #89? What cases specifically?

**Comment by CSurieux** [comment]
Well...let's take the actual implantation of tags.  The controller build a shape and pass it to an mvc view. To add a title to my page and have it displayed in a zone different from header was not easy. I found a solution, certainly not the best, adding a new shape from my view using .New and adding it to the zone in the layout. This would have been better if from the controller there has been for example a List content item created, I could have benefit of placement and all the alternates features without a tricky feature in the view phase of MVC concept. EventuallytheListcontent item, in its 'not stored' version could be adapted. There could be an adaptation file which says for tags add dynamically this part and remove this one, something as the placement .info file, a dynamic composition file.

**Comment by bleroy** [comment]
Yeah, no, that can't work, and it's completely at odds with the way thing are done in Orchard. Content items, dynamic or not, have nothing to do with this. In the scenario you describe, no matter what, you'll have to write code somewhere (you did it in the view, not the best choice, but it works), to mutate the shape or add new ones to the tree. Creating a new shape and adding it to the zone you want is about as simple as it can get.
But the scenario that you've been asking for before is not only different, it's already possible (but definitely not what you want).

**Comment by CSurieux** [comment]
Ha Ok, I was very proud of this concept, especially the dynamic composition file which would have allowed different contentitem composition according Placement.

**Comment by bleroy** [comment]
Right, I think we agree that placement should work for everything, and that it should be one of several ways to place shapes on the page.

**Comment by CSurieux** [comment]
I am not sure you read what I tried to explain ... certainly badly.
Actually Orchard has been using contentPart and ContentItems to display data.
But with Localization and recently with LayoutPart, parts were used for purely system and technical purpose.

Why not officialize this with 'technical parts' which could be welded into some dynamically built technical content item to facilitate renderring of routes which have no direct content target ?
The main advantage is this process could leverage all the existing infra built for content to render any request.

This shoudl be the purpose of these dynamic and NO DB STORED contentitems, because built one shoot on the fly.

One obvious dynamic NoDB contetnItem will be the List dynamic part, aggregated with a dynamic title part and other parts maybe directly in controller, may be in the display filters it will easilly render the list of tags I was evoking previously. 

I agree we could quite do this with existing DB parts in Orchard 1, but it is simply a trick, nothing is here to enforce it.

Last point, using placement you could also tell which parts should be aggregated for a specific dynamically named contentitem, not only which parts should not be displayed. 
Such an engine could allow aggregating different parts depending on the actual Match for the request, which differs highly from the statically defined aggregations of the same parts read from the DB.

**Comment by bleroy** [comment]
I'm sorry if I misunderstood, but your English is sometimes hard to follow (I know where you're coming from).

No, parts and items are not at all about displaying data. They are (or at least, should be) pure model and should not in principle bleed into the view (they do in fact, as it's customary to expose the content type on shapes, for pragmatic reasons, but one should refrain from using that as much as possible).

The structures used to display data are shapes, which should be considered pure view models, and are independent, but may duplicate, from the parts and items.

Controllers that are not displaying contents, in Orchard, are still commonly returning shapes, the very same structures that part drivers emit. They are dynamic, and are not tied to the DB.

In other words, the structure you're asking for already exists, and is already widely used. They are *exactly* shapes.

On a slightly different topic, no, the scenario that I explained about, which is to build a content item dynamically, without necessary ties to the database, is one we explicitly considered when designing the content type system. Our idea was that not all contents would necessarily come from the database (think web services, etc.). That's an idea that is taken further in Orchard 2 with storage providers.

Finally, placement is already designed not only to show and hide shapes, but rather to dispatch shapes to arbitrary zones (which are also shapes). This could be generalized easily.

**Comment by CSurieux** [comment]
Ok, let's be fair with the orchard bashing, I must admit that I like maintaining this European facet in my English.

Anyway when a controller, not handling a specific content, returns a shape, I would appreciate to be able to add, to continue with the Tags example, a TitlePart and to place this TitlePart in the zone I prefer. Actually the code of the controller fills the shape and there are very few options to change it without writing code. And the content ends by default in the Layout Content Zone.
With a dynamic composition file, it could be easy to say that I add a temporary TitlePart when the request matches a contenttype name (dynamically fixed in the controller, may be the shape name ?) or a Path or a DisplayType ( I would also add a Layer ).
I could also say that the dynamic results goes in Zone different from Content.

**Comment by bleroy** [comment]
You can add multiple shapes, exactly like part drivers can. You can add a title *shape*, and a tags *shape* and place those using code. You get to choose what zone you send them to, it doesn't have to be `Content`. I agree that having placement apply to this kind of shape, not just the shapes emitted from part drivers, would be nice. In fact, I agree so much that it's exactly what I did in DecentCMS a year ago ;)

Again, shapes already do exactly what you're asking for, so you can stop insisting about adding parts into a context where they don't belong.

**Comment by CSurieux** [comment]
In all the elegant models there is a time where we must find a solution to converge and do something.
The fact that I name this Part is shocking your orchard father and technical architect of Orchard.
What I try to explain is that I started from those CoreShapes (List, Zone, Pager, etc.) and wanted to introduce better control on their rendering.

Then I realized that the process of going from the ContentItem through Drivers, Filters, etc. to Rendering is far better formalized, rich and easy to use than the Shapes. From this analogy I was incited to consider that these 'extended CoreShapes' could be flowed by code as 'dynamic Parts'.

But if this is totally in contradiction with Newton Laws version Orchard....Let's say that having some way to better control, and increase variety, of Core Shapes should be enforced.

**Comment by bleroy** [comment]
Part has a specific meaning in Orchard since day 1. Deal with it. I'm done with this conversation as you are clearly more interested in name calling and spite than in a meaningful technical conversation. You're not even listening to what I'm explaining, even when I'm actually agreeing with you (or the other way around).

**Comment by CSurieux** [comment]
No, No I carefuly read and understand your points. And agree with 80%. Not your case. 
Should discussion stay at the meanigful technical level ?  
Sebastien and you are doing a reingeneering of your first Orchard version, and to do this you must start with a fresh mind, removing certitudes from your first work. This is not obvious and one of the classic pitfall of reingeneering  when done by the original authors. It could be a very positive fact but it has drawbacks.
It appears that the V 2 will mainly be rewritten by you 2, which may be a MS requirement. The others could have questions. Questions as which level of reingeneering. Little by little we discover the direction. 
Some open source communities could adopted a more structured way.
May be you are right and considering the status of Orchard, a rewritte driven by tdev masters having done the V1 is the best solution ?
